# 3. Introduksjon til programmering

### Biblioteker til oppgavene

In [6]:
!install --upgrade pip3

install: illegal option -- -
usage: install [-bCcpSsv] [-B suffix] [-f flags] [-g group] [-m mode]
               [-o owner] file1 file2
       install [-bCcpSsv] [-B suffix] [-f flags] [-g group] [-m mode]
               [-o owner] file1 ... fileN directory
       install -d [-v] [-g group] [-m mode] [-o owner] directory ...


In [8]:
%pip install numpy

Defaulting to user installation because normal site-packages is not writeable
  Using cached numpy-1.23.2.tar.gz (10.7 MB)
  Installing build dependencies ... done
  Getting requirements to build wheel ... done
    Preparing wheel metadata ... done
  ERROR: Command errored out with exit status 1:
   command: /Library/Developer/CommandLineTools/usr/bin/python3 /Library/Developer/CommandLineTools/Library/Frameworks/Python3.framework/Versions/3.8/lib/python3.8/site-packages/pip/_vendor/pep517/_in_process.py build_wheel /var/folders/0h/pp5g9p3x6yb_j3l0x6s_hz4r0000gn/T/tmp2tlra_1u
       cwd: /private/var/folders/0h/pp5g9p3x6yb_j3l0x6s_hz4r0000gn/T/pip-install-zqyom_51/numpy
  Complete output (1641 lines):
  Running from numpy source directory.
  setup.py:86: DeprecationWarning:
  
    `numpy.distutils` is deprecated since NumPy 1.23.0, as a result
    of the deprecation of `distutils` itself. It will be removed for
    Python >= 3.12. For older Python versions it will remain present.
    I

In [1]:
import numpy as np

ModuleNotFoundError: No module named 'numpy'

## 3.1.1 Input og output
### Oppgave 3.1
Lag et program som skriver ut “Hei, navnet ditt”. Modifiser så programmet slik at navnet ditt lagres i en variabel, og benytt så denne variabelen i utskriftskommandoen.

#### Løsning:

### Oppgave 3.2
Lag tre variabler av ulik type og gi dem en verdi. Skriv ut de tre verdiene.

#### Løsning:

### Oppgave 3.3
Lag et program som skriver ut tre variabler av ulik type etter hverandre.

#### Løsning:

### Oppgave 3.4
Lag et program som skriver “Hei,navn!” til skjermen. Brukeren skal bes om å taste inn navnet sitt. Dette navnet skal lagres i en variabel, og så skrives ut som i setningen ovenfor. Husk også utropstegnet til slutt.

#### Løsning:

<b>Utvid</b> programmet til å spørre om alderen din. La programmet skrive ut en kommentar til alderen din til slutt, for eksempel “din alder er jammen gammelt!”

#### Løsning:

### Oppgave 3.5
Lag et program som spør etter to tall fra brukeren. Programmet skal så skrive ut disse to tallene. Lag en passende utskriftssetning.

#### Løsning:

## 3.1.2 Matematiske operasjoner
### Oppgave 3.6
Lag to variabler med hver sin tallverdi. Lag en tredje variabel med summen av disse. Skriv ut summen slik: “Summen av tall 1 og tall 2 er sum”. 

#### Løsning:

<b>Modifiser</b> programmet slik at det tar tallene fra brukeren.

#### Løsning:

### Oppgave 3.7
Forklar hva som er forskjellen mellom kodene

```
    print(a/(b*c))
```

og
```

    print(a/b*c)
```

#### Løsning:

Svar her

### Oppgave 3.8
Lag et program som regner ut forholdet mellom massen til protonet, mp, og massen til elektronet me. Programmet ditt skal gi at forholdet blir omtrentlig lik $1836$. Du kan bruke at $mp = 1.672621898 · 10^{−27}$ og $me = 9.10938356 · 10^{−31}$.
#### Løsning:

### Oppgave 3.9
Lag et program som regner ut gjennomsnittet av et visst antall tall.
#### Løsning:

### Oppgave 3.10
Hvert program i denne oppgava inneholder noen feil. Finn feilene
og rett på dem slik at programmene kan kjøres korrekt. 

<b>1.</b> Programmet skal regne ut sinus av $30º$:

```Python
a = sin(30)
print(f'Sinus av 30° er = {a}')
```

#### Løsning:

<b>2.</b> Programmet skal regne ut prisen på 3 kaffe- og 4 tepakker:

```Python
pris_kaffe = 45 kr
pris_te = 40 kr
totalpris = 3*pris_kaffe + 4*pris_te
print(f'Total pris: {totalpris}')
```

#### Løsning:

<b>3.</b> Programmet skal regne ut produktet av a og b, som begge er tall gitt fra brukeren:

```Python
c = a * b
a = input('Skriv inn verdi for a: ')
b = input('Skriv inn verdi for b: ')
print(c)
```

#### Løsning:

### Oppgave 3.12
(Biologi)
Et menneske vil dø dersom han/hun mister så mye vann som $15$ % av sin kroppsvekt. Skriv et program som ber brukeren om å skrive inn hvor mye en person veier. Programmet skal deretter regne ut hvor mange liter vann som utgjør $15$ % av personens vekt. Resultatet skal så skrives ut.

#### Løsning:

### Oppgave 3.13
(Kjemi) Lag et program som regner ut hvor mye en trenger å veie ut av et bestemt stoff for å lage et visst volum (gitt av brukeren) av en løsning med en viss konsentrasjon (gitt av brukeren).

For en kjemiker vil antall mol av et stoff være viktigere å kjenne til enn antall gram. Molariteten (M) til et stoff er antall mol av det løste stoffet per liter løsning. I SI-enheter vil konsentrasjon da være mol/dm³.

$$M=\frac {\rm {mol~lost~stoff}}{\rm {volum~av~hele~losningen}}$$

#### Løsning:


### Oppgave 3.14
Skriv et program som beregner hvor mange kroner, $p$ prosent av $k$ kroner utgjør. Programmet skal be brukeren om verdiene til $p$ og $k$.

#### Løsning:

### Oppgave 3.15
Lag et program der du kan velge mellom tre godterityper: Sur sild, melkesjokolade og potetgull (valgfri smak). La oss si at sur sild koster $16.00$ kr/hg, melkesjokolade koster $25.00$ kr/hg og potetgullet koster $20.50$ kr/hg. Spør brukeren om hvor mye penger han/hun har, og fortell brukeren hvor mye godteri han/hun får av den valgte typen.

#### Løsning:
Denne oppgave skal nok være under if-test-kapittelet 

### Oppgave 3.16
En iskrembod har nettopp begynt å selge kuleis. Boden har 3 liter is for hver smak. Hver kule som blir solgt, er en perfekt kule med radius på 2.5 cm. Lag et program som regner ut og skriver ut hvor mange kuler 3 liter is vil gi.

#### Løsning:

### Oppgave 3.17
Lag et program som regner ut den briggske logaritmen (tierlogarit- men) til et tall. Utvid programmet slik at brukeren gir tallet som input.

#### Løsning:

### Oppgave 3.18
Skriv et program som tar inn et antall år fra brukeren og konverterer det til sekunder. Du kan se bort fra skuddår.

#### Løsning:

### Oppgave 3.19
Skriv et program som ber brukeren om å skrive inn to tall $a$ og $b$. Tallet $a$ skal deles på $b$. Resultatet skal skrives ut på formen:
$a$ delt på $b$ = antall ganger a går opp i $b + rest$

#### Løsning:

Detter er åpenbart feil. Rest må deles på b

## 3.1.3 Funskjoner
### Oppgave 3.20
Hvert program i denne oppgava inneholder noen feil. Finn feilene
og rett på dem slik at programmene kan kjøres korrekt.

#### 1. 
Funksjonen skal regne ut $f(x) = \frac{1}{2x}+1$ og returnere svaret. I programmet skal funksjonen kalles på for $x = 4$:

In [ ]:
def f(x):
    1/2*x + 1

print(f'f(x) = {f(x)}')

#### 2. 
Programmet ber brukeren om å skrive en verdi for $x$ og regner ut $f(x) = 3x+cos(x)$ og returnere svaret.

In [ ]:
def f(x):
    return 3x + cos(x)

y = input("Skriv inn verdi for x: ")
print(f'f(x) = {f(y)}')

#### 3.
Programmet skal skrive ut funksjonsverdien til G.

In [ ]:
def G(t):
    a = t**2 - 2
    return a

t=1 
print(a)

### Oppgave 3.21
Lag et program eller flere programmer som regner ut: 
1. Arealet til en sirkel.
2. Radius til en sirkel gitt diameteren.
3. Omkretsen til en sirkel.
4. Volumet til ei kule.
5. Overflatearealet til ei kule.

#### Løsning:

In [ ]:
#1


#2


#3


#4


#5


### Oppgave 3.22
Forklar hvorfor programmet under gir None som output. Rett på programmet slik at x får en tallverdi fra $f$.

In [ ]:
def f(x):
    x = 2*x
    
x = 12
x = x + 12
x = f(x)
print(x)

#### Løsning:

-- Skriv svaret her --


### Oppgave 3.23
Amerikanere er rare dyr. De har både Trump og Fahrenheit. Det er vanskelig å lage en funksjon som erstatter Trump med noen andre, men vi kan lage en funksjon som konverterer Fahrenheit til Celsius (ikke motsatt – det vil vi jo ikke!). Formelen er slik:

$$T_C = (T_F-32)\cdot \frac{5}{9}$$

der $T_C$ er temperaturen i Celsius og $T_F$ er temperaturen i Fahrenheit. Lag et program som konverterer Fahrenheit til Celsius gjennom en funksjon.

#### Løsning:

### Oppgave 3.24

Karvonens formel kan brukes til å finne pulsen til en person gitt makspuls $H_{maks}$, hvilepuls $H_{hvile}$ og treningsintensitet p (i prosent):

$$\textrm{Hjerteslag per minutt} = (H_{maks}-H_{hvile})\cdot \frac{p}{100}+H_{hvile}$$

der $H_{maks}$ er maks antall hjerteslag personen kan ha. Den maksimale pulsen kan en finne ved $H_{maks} = 220 − $alder til person. Lag et program som regner ut pulsen til en $20$ år gammel person som trener med $60$% intensitet og hvilepuls på $70$ slag per minutt. Lag Karvonens formel som en funksjon.

### Oppgave 3.25
Lag et program som konverterer et vinkelmål i grader til radianer ved hjelp av en egendefinert funksjon.

#### Løsning:

### Oppgave 3.26 (Fysikk) 
Lag et program som ber brukeren om å skrive inn en avstand $d$ i meter en løper skal løpe, og hvilken fart $v$ i m/s løperen starter med.
Løperen holder samme fart over hele avstanden. Finn et uttrykk for hvor lang tid løperen vil bruke, og implementer dette i en funksjon. La programmet regne ut og skrive ut tida ved å kalle funksjonen en print-funksjon.

#### Løsning:

### Oppgave 3.27 (Kjemi) 
I kjemi har vi ofte bruk for molregning. Lag et enkelt program som regner ut antall mol dersom brukeren taster inn molmasse og masse av et bestemt stoff. Du kan også be brukeren taste inn stoffet det gjelder, slik at du får dette som output også. Lag formelen som en funksjon.

#### Løsning:

### Oppgave 3.28 (Kjemi)
Lag et program som regner ut pH fra $[H_3 O^+]$ ved hjelp av en funksjon.

### Oppgave 3.29 (Fysikk)
Bruk Bohrs formel for spektrene til hydrogenatomet. Lag et program som regner ut bølgelengden til fotonet som emitteres når et elektron deeksiterer fra skall m til n. Bruk en funksjon.

#### Løsning:

### Oppgave 3.30 (Fysikk)
Programmer én av bevegelsesformlene som en funksjon. Du kan selv velge hva programmet regner ut.

#### Løsning:

   ### Oppgave 3.31 
Lag en funksjon som tar tre variabler a, b og c, tilsvarende koeffisientene i andregradsfunksjoner av typen $f(x)=ax^2+bx+c$. La funksjonen løse andregradslikninger av typen $f(x)=0$ ved hjelp av andregradsformelen.

#### Løsning:

### Oppgave 3.32
Hvorfor har x samme verdi før og etter funksjonen f har blitt kalt i programmet under?

In [ ]:
def f(x): 
    x=x+3
    return 9*x

x=3
print(x) # Skriver ut 3
y = f(x)
print(x) # Skriver ut 3

#### Løsning:

-- Skriv svaret her --

### Oppgave 3.33
Den briggiske logaritmen til en verdi x forteller oss hva tallet 10 må opphøyes i for å få verdien x. I denne oppgava skal vi se på andre logaritmer.
### a)
I informatikk er det ofte nyttig å finne hvor mye tallet 2 må opphøyes i for å gi x som svar. Dette skriver en gjerne $\log_2(x)$. For å regne ut $\log_2(x)$, er det mulig å bruke den briggske logaritmen:

$$\log_2(x)=\frac{\log(x)}{\log(2)}$$

Skriv et program som ber brukeren om en verdi x, regner ut $\log_2(x)$ og skriver ut resultatet. Du kan f.eks sjekke om programmet ditt skriver ut $8$ som svar når $x=256$.

#### Løsning:

### b)
Det er også mulig å finne ut hva et vilkårlig tall a må opphøyes opp i for å gi x som svar. Dette skriver vi gjerne $\log_a(x)$:

$$\log_a(x)=\frac{\log(x)}{\log(a)}$$

Utvid programmet ditt fra a) slik at det ber brukeren om en verdi for a og x, regner ut $\log_a(x)$ og skriver ut resultatet. Du kan f.eks sjekke om programmet ditt skriver ut $3$ som svar når $a = 5$ og $x = 125$.

### Oppgave 3.34 (Fysikk) 
De fleste gasser kan modelleres med tilstandslikninga for idelle gasser:

$$PV = nRT$$

der $P$ er trykket i pascal, $V$ er volumet i kubikkmeter, $n$ er stoffmengden $i$ mol, $R$ er gasskonstanten $8.3144598$J/(mol · K) og $T$ er temperaturen i Kelvin.

1. Lagetprogramderduløserlikningaforulikevariabler,gjernegittavbrukeren. 
2. Utfordring:Søk opp og sett deg inn i Vander Waals-likninga. Dette er en mye bedre modell for gasser, spesielt ved høyt trykk og lav temperatur. Prøv å lage et program som bruker denne likninga som modell istedenfor tilstandslikninga
for ideelle gasser.
3. Hva er forskjellen på de to modellene for gasser som du har lagd programmer
for ovenfor? Hvorfor har vi to modeller når den ene i utgangspunktet alltid gir mer nøyaktige resultater enn den andre?

#### Løsning:

In [ ]:
#1


In [ ]:
#2


In [ ]:
#3


### Oppgave 3.35
Funksjoner kan også ta inn andre funksjoner som argument. I matematikken kalles en funksjon $f$ for sammensatt dersom $f(x) = g(u(x))$. Programmet under skal regne ut $f(x)$ gitt funksjonene $u$, $g$ og verdien $x$. Erstatt alle spørsmålstegn(?) slik at programmet under vil kjøre og regne ut $f(x) = g(u(x))$.

In [ ]:
import ? as np

def g(x):
   return ?.sqrt(x)

def u(x):
   return x**2 + 7

def f(g,u,x):
   return g(u(x))

x=3
print(f'f(x) = g(u(x)) = {f(?,?,?)} for x = {x}')

#### Løsning:

### Oppgave 3.36 (Matematikk R1)
Lag et program som finner ut hvor mange kombinasjoner det er mulig å gjøre fra et utvalg av $r$ elever fra totalt $n$ elever der rekkefølgen ikke har noen betydning. Du kan bestemme selv om verdiene for $n$ og $r$ skal være bestemte verdier i koden eller bli gitt fra brukeren som input.

#### Løsning:

### Oppgave 3.37
Skriv en funksjon som tar inn ei rente i prosent, et beløp og et antall år $t$. Beløpet skal ligge i en bankkonto som argumenter. La funksjonen beregne og returnere hva beløpet blir når det tas ut fra kontoen etter $t$ år. Test om funksjonen din gir omtrent ≈ $162.89$ kroner etter at $100$ kr med $5$ % rente har ligget i bankkontoen i $10$ år.

#### Løsning: